replace non-breaking spaces with plain spaces

StefanKarpinski opened the pull request:
Also end all text files with newlines.

StefanKarpinski committed:
whitespace: end text files with single newlines

StefanKarpinski committed:
whitespace: replace non-breaking space => space

StefanKarpinski commented:
Some of my most important work.

goretkin commented:
These kinds of changes can be a bit disruptive as far as git blame, etc goes. There may be a way to mitigate it: https://tekin.co.uk/2020/09/ignore-linting-and-formatting-commits-when-running-git-blame

vtjnash commented on test/file.jl:
```diff
@@ -193,7 +193,7 @@ end
             push!(temps, t)
             @test ispath(t)
-            @test length(TEMP_CLEANUP) == i 
+            @test length(TEMP_CLEANUP) == i 
```
?

StefanKarpinski replied:
The real question is why that trailing space wasn't caught by the whitespace check...

StefanKarpinski replied:
Ah, fun. Our whitespace check only checks for trailing normal spaces.

StefanKarpinski commented:
This is mostly adding trailing newlines which don't affect git blame.

StefanKarpinski committed:
whitespace: use only UNIX line endings (\n)

StefanKarpinski commented:
Well this snowballed. We now check for all trailing whitespace, and that line endings are UNIX style (`\n` only, no `\r\n`), and that text files end with exactly one newline.

KristofferC commented:
I think it would be worth adding a git blame ignore file for this afterward.

StefanKarpinski commented:
No problem if you want to do it.

simeonschaub commented on contrib/check-whitespace.sh:
```diff
@@ -24,8 +24,7 @@ file_patterns='
 
-# TODO: Look also for trailing empty lines, and missing '\n' after the last line
-if git --no-pager grep --color -n --full-name -e ' $' -- $file_patterns; then
+if git --no-pager grep --color -n --full-name -P '[\s ]+$' -- $file_patterns; then
```
That should do the same thing, right?
```suggestion
if git --no-pager grep --color -n --full-name -P '\s+$' -- $file_patterns; then
```

StefanKarpinski replied:
It should but it doesn't—git does not consider nbsp to be a space. In general git's regex matching here is very broken. In basic or POSIX mode, it treats file contents as Latin-1, which is wrong (took me a while to debug that). In Perl-compatible mode it at least treats the text as UTF-8, so that works but it doesn't correctly. However, even in Perl-compatible mode `\s` doesn't match nbsp. So this uses the literal nbsp character in a class. Of course there are other space characters that still aren't matched, which isn't great but those seem pretty unlikely to sneak into the code base.

simeonschaub replied:
I'm sorry I asked! 😂

StefanKarpinski replied:
Of course, I could have done something more drastic like rewrite this code to use grep or perl, but that seemed harder and worse than just doing this nasty hack.

vtjnash replied:
This might be my fault. When you run code via a Makefile, we force it into `C` locale (aka latin-1), whereas if you run it directly, you are probably using `en_US.UTF-8` or something similar.

https://github.com/JuliaLang/julia/blob/03433a20a848ee1c07cfcae6283ec25e01b49a18/Make.inc#L100-L102

staticfloat replied:
Ah yes, the classic James Mickens situation: "I have no tools because I have destroyed my tools with my tools."

StefanKarpinski replied:
Ah, well, I didn't see this until now and I've rewritten the check script in Julia, which is easier to maintain and extend if we want to check for other things. Good to know that this isn't due to git being completely flaky, but rather due to UNIX and POSIX having a completely awful locale system, which was a bad idea, but at least we all know it's a bad idea.

StefanKarpinski commented:
@staticfloat, I assume that the whitespace check that's failing is running the version of `contrib/check-whitespace.sh` that's on this branch, right? If so, this failure extra fun because it is claiming that line 56 of `src/flisp/julia_opsuffs.h` has trailing whitespace. Which it doesn't: the line ends with `ᵠ`. However, the UTF-8 encoding of `ᵠ` is `e1 b5 a0` and `a0` is the Latin-1 encoding of non-breaking space (code point U+00a0). So that failure indicates that while my git's Perl-compatible mode correctly treats files as UTF-8, the git on the test machines is compiled differently and treats files as Latin-1 instead. Potential solutions:

1. Upgrade the git binary on the test machines
2. Rewrite the whitespace check using a tool that handles UTF-8 correctly

Option two seems better and more reliable. This check would be easy to implement in Julia, but we don't have Julia at that point, so presumably the check would have to be implemented in some other language like Python or Perl. What language do we already depend on that I can do this in?

staticfloat commented:
> I assume that the whitespace check that's failing is running the version of contrib/check-whitespace.sh that's on this branch, right?

Yes, that's correct.  You can see precisely which gitsha (`0d072a4be93a3df9e1b092acd6af2e1ad45e368c`) was checked out in the [`preparing work directory` section of the build output](https://buildkite.com/julialang/julia-master/builds/9302#29bc69b3-a727-4bcc-b289-91608fa7884c/67-97).

> Upgrade the git binary on the test machines

The test job is [running inside of a rootfs](https://github.com/JuliaLang/julia/blob/0d072a4be93a3df9e1b092acd6af2e1ad45e368c/.buildkite/pipelines/main/misc/whitespace.yml#L14-L18), which is installing [the default `git` from debian when we `debootstrap`](https://github.com/JuliaCI/rootfs-images/blob/454adcd22801a3d097fc7abdb1384c3b0677f645/linux/package_linux.jl#L20).  If [we use `test_rootfs.jl`](https://github.com/JuliaCI/rootfs-images/blob/main/test_rootfs.jl) to get a shell inside the sandbox, we can see what version of `git` that actually installs, and whether your suspicion is correct:

```
$ julia --project test_rootfs.jl -a x86_64 -t 2a058481b567f0e91b9aa3ce4ad4f09e6419355a -u https://github.com/JuliaCI/rootfs-images/releases/download/v4.8/package_linux.x86_64.tar.gz
[ Info: Artifact did not exist locally, downloading
[ Info: Unpacking /tmp/jl_0iuc5f-download.gz into /home/sabae/.julia/artifacts/jl_tplFGg...
juliaci@kholin:/build$ git --version
git version 2.20.1
$ git clone -b sk/nbsp https://github.com/JuliaLang/julia
Cloning into 'julia'...
remote: Enumerating objects: 365316, done.        
remote: Counting objects: 100% (15/15), done.        
remote: Compressing objects: 100% (15/15), done.        
remote: Total 365316 (delta 3), reused 5 (delta 0), pack-reused 365301        
Receiving objects: 100% (365316/365316), 224.39 MiB | 35.13 MiB/s, done.
Resolving deltas: 100% (270641/270641), done.
juliaci@kholin:/build$ cd julia
juliaci@kholin:/build/julia$ ./contrib/check-whitespace.sh 
src/flisp/julia_opsuffs.h:56:   0x00001d60, // ᵠ
src/flisp/julia_opsuffs.h:68:   0x00001da0, // ᶠ
Error: trailing whitespace found in source file(s)

This can often be fixed with:
    git rebase --whitespace=fix HEAD~1
or
    git rebase --whitespace=fix master
and then a forced push of the correct branch
```

So we could try a different built of `git` or something, but mostly I just wanted to highlight that we've built a good chunk of the functionality of `docker` on top of Artifacts, Sandbox, etc... :P

> This check would be easy to implement in Julia, but we don't have Julia at that point

Perhaps somewhat surprisingly, we can have Julia here; with one small tweak, we can have the Julia that we used to setup the sandbox (v1.6) available inside of the sandbox.  [Here's the tweak](https://github.com/JuliaLang/julia/pull/44348/files), and [here's the proof](https://buildkite.com/julialang/julia-master/builds/9347#9b37c0f0-d84b-4fd2-8d76-cf026fb662ef/159-163)

StefanKarpinski commented:
@staticfloat, what do you prefer? Should I reimplement the whitespace check as a Julia program or is it easier to upgrade `git` on the CI systems?

staticfloat commented:
It's up to you, but I'm not confident that a newer version of `git` from `apt` will work; my guess is it's some transitive dependency of `git` that needs to be altered, and that might be complicated to fix.

In addition to whatever version of Julia you desire, you have access to the following tools (or newer, if we upgrade the rootfs image):
```
$ python3 --version
Python 3.9.2
```

```
# perl --version

This is perl 5, version 28, subversion 1 (v5.28.1) built for x86_64-linux-gnu-thread-multi
(with 65 registered patches, see perl -V for more detail)
```

StefanKarpinski committed:
whitespace check: rewrite in Julia, add checks

This now checks for:

- trailing non-ASCII whitespace
- non-breaking spaces anywhere
- non-UNIX line endings
- trailing blank lines
- no trailing newline

Git can't handle this, largely because whether it interprets files as
UTF-8 or Latin-1 depends on how system libraries that it uses for regex
matching are compiled, which is inconsistent and hard to fix. The end of
file checks are also quite awkard and inefficient to implement with Git
and shell scripting. Julia is fast and lets us present results clearly.

StefanKarpinski commented:
Ok, wrote the whitespace check script in Julia.

simeonschaub commented on contrib/check-whitespace.jl:
```diff
@@ -0,0 +1,55 @@
+
+    for line in eachline(path, keep=true)
+        lineno += 1
+        contains(line, '\r')   && file_err("non-UNIX line endings")
```
Should this report the line number too? Maybe at least if it's not the first line, or do you think it's very unlikely for inconsistent newline characters to occur in the same file?

StefanKarpinski replied:
If you include the line number then you get a separate error message for every single line of the file, which is generally unhelpful. This will happen if someone edited the file on Windows in an editor that converts the line endings to CRLF, which it will do for all the lines, not just the ones that were edited. And the line number isn't helpful for fixing it either since you  generally ask your editor to convert the line endings of the entire file to CR. There aren't any editors that let you save part of a file with CRLF and other parts of a file with CR line endings.

simeonschaub replied:
Yeah, I guess that makes sense

staticfloat commented:
https://buildkite.com/julialang/julia-master/builds/9829#d496a327-7923-4420-be12-9c4777abb2b9/160-163 Huzzah!

StefanKarpinski commented:
Btw, please DON'T squash merge this—having the changes separate makes it much easier to resolve any merge conflicts that may come up since each commit does a specific thing that can be automated.

StefanKarpinski commented:
This is good as far as I'm concerned. If it passes CI it can be merged (non-squash).

KristofferC commented on .buildkite/pipelines/main/misc/signed_pipeline_test.yml.signature:
```diff
@@ -1 +1 @@
-Salted__���J0�Q?���rۀ�g�~�d��ۛŧ�ө��o���Ujʀ���p�)�$�U$���y@gZM}{�m��,۠�K��e�r�
\ No newline at end of file
+Salted__"@�:�"z,͡�Ƣ�t9���"��?�+�'u��"e��:8�Z�kg���>�� ��ȫ.�7U12M~ ������u���
```
Where these meant to be changed?

staticfloat replied:
Yeah, I blanket re-signed.

StefanKarpinski replied:
Yes: the YAML files that these signatures apply to were changed, so the signatures have to change too. Might have been better to not change them, but having all text files end with a single newline seems like a simple and consistent rule that's close to what we were already doing.

StefanKarpinski commented:
It would be ideal to get all CI passes green but it seems pretty unlikely that this change is related to the various failures here. Up to those managing these CI failures to decide what to do here (wait for green or merge before).

DilumAluthge commented:
We'll need to install Julia inside the whitespace job: https://github.com/JuliaCI/julia-buildkite/pull/58

StefanKarpinski commented:
Can you CI guys fix this up with additional commits to make it work? I can rebase the history once it's working but I don't think I can get this working myself with the CI stuff in flux.

DilumAluthge commented:
Is it cool if I push directly to this branch? And then you can rebase when it's all green.

StefanKarpinski commented:
Yeah, go for it. Just add commits and I'll fix it up later.

DilumAluthge commented:
@StefanKarpinski Looks like Buildkite is green on this now.

StefanKarpinski commented:
Does that mean it's safe to merge this PR? Buildbot failures look unrelated.

DilumAluthge commented:
👍

DilumAluthge commented:
Yes, I think you can merge this PR.
